feat: CXSPA-3794 - SSR contextual logging doc input (#1537)
SAP/spartacus-docs: _pages/dev/ssr/server-side-rendering-contextual-logging.md

Author: TL;DR section has been added. thank you for this suggestion and let's wait for @gladius-mtl opinion 👍 :)

---
title: Server-Side Contextual Logging
feature:
- name: Server-Side Contextual Logging
  spa_version: 6.2
  cx_version: n/a
---

<TODO Verify names in code examples>

Server-Side Contextual Logging is a feature that has been introduced to enhance the debugging experience for developers. With this feature, developers can see the context of the printed output and the source of the log message. Additionally, it ensures that the content is formatted correctly so that it is readable by developers, as well as monitoring tools. 

In addition to the built-in solutions, this feature also comes with tools that allow customers to customize the logging experience. 

Without contextual logging, the following issues arise:

- Log messages are just strings with little context about the rendered URL, making it difficult to connect the log message with the source of the log.

> - Log messages are not standardized, which can make it difficult to read and understand the log messages.

Reviewer: ```suggestion
- Log messages's format is not standardized, which can make it difficult to read and understand the log messages.
```

> - If the log message is multiline, it is not formatted correctly, and each line is treated as a separate log in the monitoring tools, further complicating the log viewing experience.

Reviewer: ```suggestion
- If the log message is multiline, it is not formatted correctly, and each line is treated as a separate log in the monitoring tools, further complicating the log viewing experience. (This can be the case e.g. for multi-line stacktraces of errors.)
```

These issues make it difficult to read and understand log messages, particularly when log messages are coming from multiple parallel server-side-renderings and NodeJS servers. 

Thankfully, with Server-Side Contextual Logging, these issues are avoided, and developers can easily identify the source of log messages and understand them better.

Composable Storefront offers a default logger called `DefaultExpressServerLogger` to address common issues. This logger is used by default when the SSR Contextual Logging is enabled. 

> This logger takes care of proper formatting and recognizes whether the output should be readable for developers or monitoring tools. It not only logs the message but also provides information about the request that initiated the rendering process. 

Reviewer: ```suggestion
This logger takes care of proper formatting and recognizes whether the output should be readable for developers or monitoring tools. It not only logs the message but also provides information about the related request that initiated the rendering process. 
```

The example below shows how the logger creates logs for development mode. It produces a multiline JSON output:
```json
{
  "message": "Rendering started (/powertools-spa/en/USD/product/3788616/bt-dy-720-e)",
  "context": {
    "timestamp": "2023-06-19T23:34:37.845Z", // timestamp of the log
    "request": {
      "url": "/powertools-spa/en/USD/product/3788616/bt-dy-720-e", // URL of the request

>       "uuid": "1ef6d1d9-65c2-43c8-860e-32d0cb1bf698", // UUID of the request

Reviewer: ```suggestion
      "uuid": "1ef6d1d9-65c2-43c8-860e-32d0cb1bf698", // random UUID of the request
```

>       "timeReceived": "2023-06-19T23:34:37.843Z" // timestamp of the request

Reviewer: ```suggestion
      "timeReceived": "2023-06-19T23:34:37.843Z" // timestamp when the request was received
```

   }
 }
}
```
Here we can see an example of the log created for production purposes. It is a single line of JSON that is readable by monitoring tools:
```
{"message":"Rendering started (/powertools-spa/en/USD/product/3788616/bt-dy-720-e)","context":{"timestamp":"2023-06-19T23:34:37.845Z","request":{"url":"/powertools-spa/en/USD/product/3788616/bt-dy-720-e","uuid":"1ef6d1d9-65c2-43c8-860e-32d0cb1bf698","timeReceived":"2023-06-19T23:34:37.843Z"}}}
```
The following is the representation of the log message in the monitoring tool:
![Log Representation In Kibana](assets/images/../../../../../assets/images/ssr/contextual_logging_kibana_logs.png)

## Enabling the SSR Contextual Logging

To enable the SSR contextual logging, add the `logger` flag to the configuration object as the second parameter to the method that decorates `ngExpressEngine` in the `server.ts` file. The following is an example:

```ts
import { ngExpressEngine as engine } from '@nguniversal/express-engine';
import { NgExpressEngineDecorator } from '@spartacus/setup/ssr';

[...]

const ngExpressEngine = NgExpressEngineDecorator.get(engine, {
  //add logger flag here
  logger: true
});

[...]

```

> From now on, your application will use the `DefaultExpressServerLogger` and provide logging context for all server-side rendered requests

Reviewer: ```suggestion
From now on, your application will use the `DefaultExpressServerLogger` and provide request's context for every logged message during server-side rendering.
```

> 

Reviewer: ```suggestion

Note: in a future major release, it's planned to enable the `logger` flag by default.

```

To implement a custom logger, the developer must create a class that implements the `ExpressServerLogger` interface. Alternatively, they may extend the `DefaultExpressServerLogger` class. 

Once the custom logger class is created, it can be passed to the `logger` property in the configuration object. 

An example of how to do this is shown below:

```ts
import { ngExpressEngine as engine } from '@nguniversal/express-engine';
import { NgExpressEngineDecorator } from '@spartacus/setup/ssr';
import { CustomExpressServerLogger } from './custom-express-server-logger';

[...]

const ngExpressEngine = NgExpressEngineDecorator.get(engine, {
  // pass custom logger here
  logger: new CustomExpressServerLogger()
});

[...]
```

The following is an example of a custom logger implementation:

```ts
import { ExpressServerLogger } from '@spartacus/setup/ssr';

export class CustomExpressServerLogger implements ExpressServerLogger {

  //custom implementation of the `log` method
  log(message: string, context: ExpressServerLoggerContext): void {
    console.log({ message: `[My Custom LOG]: ${message}`, ...context });
  }

>   [...]

Reviewer: ```suggestion
  // don't forget to customize also other methods like: `debug`,`info`, `warn` and `error` 
  [...]
```

}
```

> An important aspect of our solution is its extensibility. We have designed our solution to be compatible with enterprise logging libraries used for logging in platform-server such as [Winston](https://www.npmjs.com/package/winston) or [Pino](https://www.npmjs.com/package/pino), making it easy to add their features to our system.

Reviewer: Let's divide the following section into 2 subsections with 2 headers:
- Plugging third party logger
- Providing additional context to logs

so the separate the examples of the customization usecases

> The following is an example of how to implement the custom logger with the `Pino` library:

Reviewer: Let's move the disclaimer from line 157 to here:

`
**Disclaimer: This is only an example of how to use a third-party logger library. Do not treat this as a recommendation of the particular library.**
`

```ts
import { DefaultExpressServerLogger, ExpressServerLoggerContext } from '@spartacus/setup/ssr';
import { pino } from 'pino';

class CustomLogger extends DefaultExpressServerLogger {

>   private logger = pino({ customLevels: { log: 15 }, level: 'log' }); // as there is no `log` level in pino, we need to define it here;

Reviewer: I've realized only now that (according to NodeJS docs) `log` [is just an alias](https://nodejs.org/api/console.html#consoleinfodata-args) for `info`. See also [an article](https://inspirnathan.com/posts/114-console-log-levels/) showing that Chrome also treats `log` and `info` on the same level they have the same logging level `"Info"`. So in our example with `Pino`, I believe we should not create a custom logger level, but just map our methods `log` and `info` to pino's `"info"` logger level.
What do you think?

Author: I agree.

  [...]

  // custom implementation of the `log` method
  log(message: string, context: ExpressServerLoggerContext): void {
    // custom log level is used here
    this.logger.log(this.handleContext(context), message);
  }

  warn(message: string, context: ExpressServerLoggerContext): void {
    //built-in log level is used here
    this.logger.warn(this.handleContext(context), message);
  }

  [...]

  // method to handle the context before it is passed to the logger
  protected handleContext(context: ExpressServerLoggerContext) {
    const { request, ...restOfContext } = context;
    // `mapRequest` is used to map the request to the log message. This is a method provided in the `DefaultExpressServerLogger` class.
    // NOTE: for some cases request object is not passed to the context, so we need to check if it exists
    const mappedRequest = request ? this.mapRequest(request) : undefined;
    return { ...restOfContext, request: mappedRequest };

>   }

Reviewer: 1. First of all, this code looks very complex, lots of comments about gotchas. TBH as a customer, I'd love to have something like this:
1) If I'm already extending `DefaultExpressServerLogger`, I shouldn't have to write 6 methods(!) (log, info, ..., error) and some context mapper just in order to add some additional context.
2) It should suffice that I override 1 method to reuse the default mapper and populate to the final context some small addiditional portion of data.

1. > // NOTE: for some cases request object is not passed to the context, so we need to check if it exists
It looks bad that customers need to know gotchas for just a simple customization use case. Maybe we made a mistake by passing ssrOptions to this concrete logger without `request`, even though it's a required property of the interface. 

1. Let's present here some very simple customization use case from real life - add to context a value of some HTTP header.

Author: Thank you for pointing this out! Yes, during writing an example of how to extend the solution provided by us, I realised it's not that straightforward and needs to be simplified. From Customer's POV provided example may look complicated showing some weak (?) points of our code. 

Fix for such an issue has been addressed in the following PR: https://github.com/SAP/spartacus/pull/17578

}
```
This is an example of how logs created by `pino` library look like:
```
{"level":15,"time":1687528211674,"pid":40850,"hostname":"SHMABCD1234567","options":{"concurrency":10,"timeout":3000,"forcedSsrTimeout":60000,"maxRenderTime":300000,"reuseCurrentRendering":true,"debug":false,"logger":"CustomLogger"},"msg":"[spartacus] SSR optimization engine initialized"}

{"level":15,"time":1687528382249,"pid":40850,"hostname":"SHMABCD1234567","request":{"url":"/electronics-spa/en/USD/","uuid":"446d5d59-0111-471e-b6f3-883617a9aaba","timeReceived":"2023-06-23T13:53:02.247Z"},"msg":"Rendering started (/electronics-spa/en/USD/)"}
```

> For more information, see: https://www.npmjs.com/package/pino

Reviewer: ```suggestion
Note: Pino library [translates log levels to integer values](https://github.com/pinojs/pino/blob/master/docs/api.md#loggerlevel-string-gettersetter). For more information, see: https://www.npmjs.com/package/pino
```

Note: This is only an example of how to use a third-party logger library. Do not treat this as a recommendation of the particular library.

**Caution:** The `DefaultExpressServerLogger` doesn't include any sensitive data in the logs by default. However, when using a custom logger, developers must exercise caution and be mindful of the data being provided. It's their responsibility to ensure that no sensitive information is being used or exposed.

> ## Using the LoggerService

Reviewer: We should strongly emphasise in this doc, that customers should use `LoggerService` instead of the global `console` if they want to automatically add request's context to their custom logs. And also any 3rd party libraries that print to the global console, that they use in their project, should also call LoggerService rather than the global `console` object.

--
Btw. We could also extract doc for LoggerService deserves into a separate standalone document. It can be mentioned in this "SSR contextual logs" doc (and vice-versa).

Author: As we discussed, for now, let's keep this information in the section about SSR. We can discuss moving the `LoggerService` description later, e.g. before the next major release.

  ]
})
export class SpartacusModule {}
```

By importing `ErrorHandlingModule`, the `CxErrorHandler` then handles all errors that occur during server-side rendering, and these errors are logged with an appropriate context.

**Note:** Since user applications may contain their own implementations of `ErrorHandler`, the `CxErrorHandler` is not used by default when contextual logging is enabled. To use the `CxErrorHandler`, it is recommended that you extend your error handler to use the `LoggerService`.

## LoggerService and Third-party Party Libraries

To ensure that you have context and proper formatting for the logs that are output by third-party libraries in your application, it is recommended that you verify whether custom loggers can be provided, and that you use the `LoggerService` if possible. Otherwise, the logs from third-party libraries will be written in plain text, without the request's context and without proper formatting.

The `LoggerService` can be used not only to log messages that are added to your applications, but also to log messages that are added to those third-party libraries that are used in your project and that communicate by their state.

> 

Reviewer 2: I tried to adjust the wording of this sentence, but I may have unintentionally changed the meaning! Can you check if it makes sense, or maybe needs rewriting a bit more? Thank you 🙂

Author: In my opinion, it makes sense :) thank you!

The following is an example of a custom logger implementation that extends the `DefaultExpressServerLogger` class and adds more functionality to the built-in `mapContext` method:

```ts
import { DefaultExpressServerLogger, ExpressServerLoggerContext } from '@spartacus/setup/ssr';

export class CustomExpressServerLogger extends DefaultExpressServerLogger {
  //extend output context with traceparent header
  override protected mapContext(context: ExpressServerLoggerContext): Record<string,any> {
     const outputContext = super.mapContext(context);
     return {
      ...super.mapContext(context),
      request: {
        // 
        traceparent: context.request.get('traceparent')
      }
     }
  }
}
```

### Plugin Third-party Logger

**Disclaimer: This is only an example of how to use a third-party logger library. Do not treat this as a recommendation of the particular library.**

The following is an example of how to implement the custom logger with the `Pino` library:

```ts
import { DefaultExpressServerLogger, ExpressServerLoggerContext } from '@spartacus/setup/ssr';
import { pino } from 'pino';

class CustomPinoLogger extends DefaultExpressServerLogger {

>   const logger = pino({

Reviewer: `const`? 👀 
```suggestion
  protected logger = pino({
```

Author: 🤦

    // custom configuration of the pino logger
  });

  // custom implementation of the `log` method
  log(message: string, context: ExpressServerLoggerContext): void {
    // Because in the console.info() function is an alias for console.log(), we can use the same method 'logger.info() for both log and info. See more: https://nodejs.org/api/console.html#consoleinfodata-args
    this.logger.info(this.mapContext(context), message);
  }

>   info(message: string, context: ExpressServerLoggerContext): void {
>     this.logger.info(this.mapContext(context), message);
>   }

Reviewer: ```suggestion
  info(message: string, context: ExpressServerLoggerContext): void {
    this.logger.info(this.mapContext(context), message);
  }
  
  // don't forget to customize also other methods like: `debug`,`info`, `warn` and `error`
```

  cx_version: n/a
---

> TL;DR How to fully enable the feature:

Author: `query:` Is the 'TL;DR' section forbidden? As we discussed with @Platonn, the document is quite long and from a customer perspective, would be great to have such (or a similar) section at the beginning of this doc that describes all steps that are necessary to fully enable contextual logging.

Reviewer: IMHO it's a feature that every customer would like to start using from day 1.
But due to backward compatibility, the process of enabling this feature requires some manual steps.
So I think we should place all those steps somewhere at the very beginning of this document, so customers can start using it immediately.

In a future major release, some of those steps will implicit defaults, then this section will be much smaller. But then I think we will still keep a disclaimer with the meaning: "how not to break this feature: use LoggerService yourself, and make your 3rd party dependencies also use LoggerService - instead of native console"

}
```

> ### Implementing Custom Loggers with Third-Party Libraries

Reviewer 2: Integrating with Third-Party Logger Libraries

export class SpartacusModule {}
```

By importing `ErrorHandlingModule`, the `CxErrorHandler` then handles all errors that occur during server-side rendering, and these errors are logged with an appropriate context.

> **Note:** Since user applications may contain their own implementations of `ErrorHandler`, the `CxErrorHandler` is not used by default when contextual logging is enabled. To use the `CxErrorHandler`, it is recommended that you extend your error handler to use the `LoggerService`.

Author: I changed a bit the meaning of this sentence.
`To use CxErrorHandler, it is recommended that you extend your handler to use the LoggerService`
Actually, we want to tell the users if they want to have context attached to errors, they should use `LoggerService` i their custom implementation of error handlers.

An example:
```
**Note:** Since user applications may contain their own implementations of `ErrorHandler`, the `CxErrorHandler` is not used by default when contextual logging is enabled. To enhance errors with the context, it is recommended that you extend your error handler to use the `LoggerService`.
```

Reviewer 2: Done.

## LoggerService and Third-party Party Libraries

To ensure that you have context and proper formatting for the logs that are output by third-party libraries in your application, it is recommended that you verify whether custom loggers can be provided, and that you use the `LoggerService` if possible. Otherwise, the logs from third-party libraries will be written in plain text, without the request's context and without proper formatting.

The `LoggerService` can be used not only to log messages that are added to your applications, but also to log messages that are added to those third-party libraries that are used in your project and that communicate by their state.

> For example, Spartacus uses the `LoggerService` to display logs created by the `i18next` library. This was achieved by creating a compatible plugin with the `i18next` configuration and using the capabilities provided by the Spartacus logging service.

Author: `query: ` Is it ok to use `Spartacus` in the new docs? 😄

Reviewer 2: Yes, because my conversion script takes care of converting these to a product name variable that is set to "composable storefront" in dita cms -- whereas if we leave them as simply the string "composable storefront", the script in its current form doesn't convert these to the product variable in dita cms. I should obviously update my script, but this is the first time I've encountered this scenario 😄
